### PR TITLE
Cartesian teleop

### PR DESCRIPTION
**Pull request by PeterMitrano**
Lets you manually jog XYZ using a gamepad. Based on the jacobian follower. Only tested with val, in sim, with a logitech controller

### CODE REVIEW

File: arm_robots/scripts/cartesian_teleop.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    def get_current_positions(self):
+        target_positions = []
+        for tool_name in self.tool_names:
+            current_tool_pose = self.robot.get_link_pose(tool_name)
+            current_tool_position = ros_numpy.numpify(current_tool_pose.position)
+            target_positions.append(current_tool_position)
+        return target_positions
+
+    def buttons_thread_func(self):
+        while True:
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:57]: `while not self.should_disconnect:`
Remove `if self.should_disconnect` later

File: arm_robots/scripts/cartesian_teleop.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            events = get_gamepad()
+            for event in events:
+                if self.gamepad.x_clicked(event):
+                    if self.current_tool_idx == len(self.tool_names) - 1:
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:61]: Personal preference, but avoid `if else` as much as possible
```
self.current_tool_idx = (self.current_tool_idx + 1) % len(self.tool_names)
```

or

```
self.current_tool_idx += 1
if self.current_tool_idx == len(self.tool_names):
  self.current_tool_idx = 0
```
```

File: arm_robots/scripts/cartesian_teleop.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        self.buttons_thread.start()
+
+        r = rospy.Rate(10)
+        while not rospy.is_shutdown():
+            self.send_commands()
+            r.sleep()
+
+        self.should_disconnect = True
+
+    def get_current_positions(self):
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:48]: `get_tool_positions` or `get_current_tool_positions` is more clear

File: arm_robots/scripts/cartesian_teleop.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                if self.gamepad.x_clicked(event):
+                    if self.current_tool_idx == len(self.tool_names) - 1:
+                        self.current_tool_idx = 0
+                    else:
+                        self.current_tool_idx += 1
+            if self.should_disconnect:
+                break
+
+    def send_commands(self):
+        delta_position = self.gamepad.get_3d_delta()
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:69]: Not thread safe - getting current tool idx vs getting 3d delta. 
Probably not worth fixing though, who cares if you accidentally move a joint a little bit

File: arm_robots/scripts/cartesian_teleop.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        target_positions = self.get_current_positions()
+        target_positions[self.current_tool_idx] += delta_position * self.delta_scale_factor
+
+        tool_name = self.tool_names[self.current_tool_idx]
+
+        target_position = target_positions[self.current_tool_idx]
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:75]: Could clean up this code

File: arm_robots/scripts/cartesian_teleop.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        rospy.logdebug(f"Controlling tool {tool_name}, target position: {target_position}")
+        target_positions = np.expand_dims(target_positions, axis=1)
+        self.robot.follow_jacobian_to_position(self.group_name, self.tool_names, target_positions, vel_scaling=1.0)
```
Comment (bsaund) [inline arm_robots/scripts/cartesian_teleop.py:78]: I don't remember how `follow_jacobian_to_position` works, but the name implies that calling this function starts an async process that uses a controller to update the robot position over time. Is it a problem that this code is repeatedly calling `follow_jacobian_to_position`, potentially before it has finished?

**Review by bsaund — APPROVED** [review]
Seems convenient!

File: arm_robots/scripts/cartesian_teleop.py (hunk 2 above)
Comment (PeterMitrano) [reply arm_robots/scripts/cartesian_teleop.py:61]: I like the later

File: arm_robots/scripts/cartesian_teleop.py (hunk 4 above)
Comment (PeterMitrano) [reply arm_robots/scripts/cartesian_teleop.py:69]: adding a lock for access to `current_tool_idx` is easy. Are there any other data races?

File: arm_robots/scripts/cartesian_teleop.py (hunk 6 above)
Comment (PeterMitrano) [reply arm_robots/scripts/cartesian_teleop.py:78]: It's blocking by default, but could be async, depending on the `blocking` variable in the robot class. If it's called repeatedly the new goal should preempt and the old goal should be canceled so it should be fine. Never tested that though I always leave it in blocking mode.

**Pull request merged**